micrometer-core/src/main/java/io/micrometer/core/instrument/binder/hystrix/MicrometerMetricsPublisherCommand.java

```diff
@@ -133,64 +103,23 @@ public void initialize() {
                 for (HystrixEventType hystrixEventType : HystrixEventType.values()) {
                     int count = hystrixCommandCompletion.getEventCounts().getCount(hystrixEventType);
                     if (count > 0) {
-                        switch (hystrixEventType) {
-                            /* this list is derived from {@link HystrixCommandMetrics.HealthCounts.plus} */
-                            case FAILURE:
-                            case TIMEOUT:
-                            case THREAD_POOL_REJECTED:
-                            case SEMAPHORE_REJECTED:
-                                Counter.builder(NAME_HYSTRIX_ERRORS)
-                                    .tags(tags)
-                                    .register(meterRegistry)
-                                    .increment(count);
-                            case SUCCESS:
-                                Counter.builder(NAME_HYSTRIX_REQUESTS)
-                                    .tags(tags)
-                                    .register(meterRegistry)
-                                    .increment(count);
-
-                                break;
-                        }
-
-                        if (executionEvents.contains(hystrixEventType)) {
-                            getExecutionCounter(hystrixEventType).increment(count);
-                        } else if (fallbackEvents.contains(hystrixEventType)) {
-                            getFallbackCounter(hystrixEventType).increment(count);
-                        } else {
-                            getOtherExecutionCounter(hystrixEventType).increment(count);
-                        }
+                        eventCounters.get(hystrixEventType).increment(count);
```

Comment:
You can benchmark it if you'd like, but just registering the counter here creates one if it doesn't exist, retrieves an existing one of it does. The cost of doing the map lookup in `eventCounters` is comparable then to the cost of calling `register`.

Reply:
I'm creating the counters at be beginning of the method anyway as it is a good practice to initialize all metrics for a command with zero to ease the aggregations/graphs afterwards.

This was already in the previous code. I update the PR to also keep the comment that was there in the previous version.

Therefore I have the meter already at hand, therefore I think it's good to use it. It also avoid to re-create the list of tags including the toLowerCase() operations. 

Therefore I'd rather keep it in the hashmap.

Reply:
Great. Thanks.

Reply:
I like how this simplifies the metric incrementing and also that they are all still set to zero initially. 

I don't feel comfortable about changing the metric breakdown though. `hystrix.execution` represented more or less how you want the 'terminal' tag to appear in that the tags all represented parts of the whole.  This change will allow duplicate counts and the user would need to know the specific tag to exclude on. I would feel better if those non-terminal executions were counted separately, under a separate name.  

We found that users of our Prometheus/Grafana dashboard got buried in the options and wanted to be able to answer specific questions:

  1. Are any of my circuits open? 
      - Un-changed
  1. Is my Hystrix code being exercised (as in is the path being used) 
      - Also unchanged, success indicates that path is being used
  2. Are fallbacks being triggered? 
      - This is a little more difficult, But achievable for Prometheus, using the regex selector on tags, but I don't want to for users to look for it. Also I don't think filtering in the same way is as easy for other backends.
  3. If they are being triggered, what is the outcome of the fallback being triggered? 
      - Same as the last point, but it is interested in filtering and displaying the results just for fallbacks

As I've mentioned before, my Hystrix experience is still pretty minimal. And the questions my dashboard were addressing probably reflect that (essentially the 'is it configured to work?' data)

Perhaps most Hystrix users would want to answer the question of 'How was my request handled in the end?' That matches with the semantics around it being `terminal`. Is it correct to say that question was impossible to answer before (or at least much more difficult)?

Reply:
@checketts A grafana dash export of your hystrix setup would go far in explanatory power. Would you mind dropping that somewhere?

Reply:
@checketts thank you for your extensive comment. It took my some days to follow up due to ongoing work projects and conferences. 

I write here to find out what a counter about fallbacks could contain. The docs are here: https://github.com/Netflix/Hystrix/wiki/Metrics-and-Monitoring#hystrix-event-types

You phrased the user's question as: "Are fallbacks being triggered"
Looking at all events that start with "fallback":
* fallback_emit -> non-terminal, should be ignored -> "no"
* fallback_success -> the fallback was triggered, and successful) -> "yes"
* fallback_failure -> the fallback was triggered, but was unsuccessful -> "yes"
* fallback_rejection -> the fallback was not attempted due to capacity problems -> "no" in a "triggered sense"? Or "yes" anyway?
* fallback_missing -> fallback doesn't exist -> "no" 

If you could help to agree on the yes/no we could wrap that into a counter.

IMHO the Hystrix events break things down into tiny pieces that are non-trivial to put together. The result depends on the question asked.

My personal conclusion with Hystrix events: the only question that can be answered straight forward is the following: how do the events relate to the total number of commands I sent. This will give you a breakdown like: 
* 50% SUCCESS (terminal)
* 50% FAILURE (non-terminal)
* 25% FALLBACK_SUCCESS (terminal)
* 25% FALLBACK_FAILURE (terminal)

Again, all terminal events sum up to 100%. Adding a counter that represents the sum of all terminal events would ease the calculation.

As a prometheus query this would look like (switch irate for rate as preferred):

````
irate(hystrix_command_event_total [1m])
/ ignoring(event) group_left
sum(irate(hystrix_command_event_total{terminal='true'} [1m]))
  by (command_group, command_name, instance, job, terminal)
````

If you want to ignore the events that never happened, add the following: 

````
... AND hystrix_command_event_total > 0 
````

Again, adding a counter that represents the sum of all terminal events would ease the calculation. I propose it to be called "hystrix_command_event_terminal_total"

Reply:
@checketts - Maybe we can take this step by step. Let me know where we agree and I will split this into two PRs. We can also have a face-to-face (skype/hangout?) for the second part.

Reply:
@ahus1 @checketts. Or if this PR represents an improvement over the status quo we can merge and iterate further?

Reply:
To make voting easier: I could rewrite it to keep all old metrics, and just add new metrics/tags. Some metrics will be deprecated (noted in the description). A switch (default off) will allow to skip all deprecated metrics. This will allow easy migration for existing users. This switch will require some documentation.